File: paper/theory-of-standing/reducers/python/reducer.py
#!/usr/bin/env python3
"""Independent Python reducer for the proof-history interchange format."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, NoReturn

INPUT_FORMAT = "theory-of-standing.proof-history.v1"
RESULT_FORMAT = "theory-of-standing.proof-result.v1"
REJECTION_FORMAT = "theory-of-standing.proof-rejection.v1"
STANDING = {"accepted", "unassessed", "superseded", "retracted"}
MAX_NAT = 9_007_199_254_740_991


class FormatError(Exception):
    pass


def fail(message: str) -> NoReturn:
    raise FormatError(message)


def exact(value: Any, keys: set[str], where: str) -> dict[str, Any]:
    if not isinstance(value, dict) or set(value) != keys:
        fail(f"{where}: expected exact keys {sorted(keys)}")
    return value


def nat(value: Any, where: str) -> int:
    if (
        isinstance(value, bool)
        or not isinstance(value, int)
        or value < 0
        or value > MAX_NAT
    ):
        fail(f"{where}: expected nonnegative safe integer")
    return value


def nat_map(value: Any, where: str) -> dict[int, int]:
    if not isinstance(value, dict):
        fail(f"{where}: expected object")
    result = {}
    for key, item in value.items():
        if not isinstance(key, str) or not key.isascii() or not key.isdecimal():
            fail(f"{where}: invalid resource id")
        resource = int(key)
        if str(resource) != key:
            fail(f"{where}: noncanonical resource id")
        result[resource] = nat(item, f"{where}.{key}")
    return result


def validate_action(value: Any, where: str) -> dict[str, Any]:
    if not isinstance(value, dict) or not isinstance(value.get("kind"), str):
        fail(f"{where}: invalid action")
    kind = value["kind"]
    if kind in {"accept", "reject"}:
        exact(value, {"kind", "claim"}, where)
        nat(value["claim"], f"{where}.claim")
    elif kind == "correct":
        exact(
            value,
            {"kind", "prior_decision", "predecessor", "replacement"},
            where,
        )
        for key in ("prior_decision", "predecessor", "replacement"):
            nat(value[key], f"{where}.{key}")
    else:
        fail(f"{where}: unsupported action")
    return value


def validate_history(value: Any) -> dict[str, Any]:
    history = exact(
        value,
        {
            "format",
            "repository",
            "authorized_performers",
            "initial_versions",
            "descriptive_dependencies",
            "records",
        },
        "history",
    )
    if history["format"] != INPUT_FORMAT:
        fail("history.format: unsupported format")
    nat(history["repository"], "history.repository")
    authorized = history["authorized_performers"]
    if not isinstance(authorized, list):
        fail("history.authorized_performers: expected array")
    for index, actor in enumerate(authorized):
        nat(actor, f"history.authorized_performers[{index}]")
    if authorized != sorted(set(authorized)):
        fail("history.authorized_performers: expected sorted unique values")
    history["initial_versions"] = nat_map(
        history["initial_versions"], "history.initial_versions"
    )
    dependencies = history["descriptive_dependencies"]
    if not isinstance(dependencies, list):
        fail("history.descriptive_dependencies: expected array")
    for index, dependency in enumerate(dependencies):
        exact(dependency, {"dependent", "depends_on"}, f"dependency[{index}]")
        nat(dependency["dependent"], f"dependency[{index}].dependent")
        nat(dependency["depends_on"], f"dependency[{index}].depends_on")
    records = history["records"]
    if not isinstance(records, list):
        fail("history.records: expected array")
    decision_ids: set[int] = set()
    for index, record in enumerate(records):
        if not isinstance(record, dict) or not isinstance(record.get("kind"), str):
            fail(f"record[{index}]: invalid record")
        kind = record["kind"]
        if kind == "submission":
            exact(
                record,
                {"kind", "claim", "producer", "scope", "authenticated"},
                f"record[{index}]",
            )
            for key in ("claim", "producer", "scope"):
                nat(record[key], f"record[{index}].{key}")
            if not isinstance(record["authenticated"], bool):
                fail(f"record[{index}].authenticated: expected boolean")
        elif kind == "verification":
            exact(
                record,
                {"kind", "claim", "scope", "property", "outcome"},
                f"record[{index}]",
            )
            for key in ("claim", "scope", "property"):
                nat(record[key], f"record[{index}].{key}")
            if record["outcome"] not in {"pass", "fail"}:
                fail(f"record[{index}].outcome: unsupported outcome")
        elif kind == "decision":
            exact(
                record,
                {
                    "kind",
                    "id",
                    "repository",
                    "authority_label",
                    "performer",
                    "expected_root",
                    "read_set",
                    "action",
                },
                f"record[{index}]",
            )
            for key in (
                "id",
                "repository",
                "authority_label",
                "performer",
                "expected_root",
            ):
                nat(record[key], f"record[{index}].{key}")
            if record["id"] in decision_ids:
                fail(f"record[{index}].id: duplicate Decision id")
            decision_ids.add(record["id"])
            record["read_set"] = nat_map(
                record["read_set"], f"record[{index}].read_set"
            )
            validate_action(record["action"], f"record[{index}].action")
        else:
            fail(f"record[{index}]: unsupported kind")
    return history


def projection(history: dict[str, Any], state: dict[str, Any]) -> list[dict[str, Any]]:
    corrected = {
        event["action"]["predecessor"]
        for event in state["events"]
        if event["action"]["kind"] == "correct"
    }
    dependencies = history["descriptive_dependencies"]
    result = []
    for claim in sorted(state["standing"]):
        needs = any(
            dependency["dependent"] == claim and dependency["depends_on"] in corrected
            for dependency in dependencies
        )
        result.append(
            {
                "claim": claim,
                "status": "needs_reassessment" if needs else "unaffected",
            }
        )
    return result


def output(history: dict[str, Any], state: dict[str, Any], code: str | None) -> bytes:
    standing = [
        {"claim": claim, "status": status}
        for claim, status in sorted(state["standing"].items())
    ]
    if any(item["status"] not in STANDING for item in standing):
        raise AssertionError("noncanonical Standing")
    events = state["events"]
    result = {
        "events": events,
        "format": REJECTION_FORMAT if code is not None else RESULT_FORMAT,
        "reassessment": projection(history, state),
        "repository": history["repository"],
        "root": state["root"],
        "standing": standing,
    }
    if code is not None:
        result["code"] = code
    return (json.dumps(result, sort_keys=True, separators=(",", ":")) + "\n").encode()


def rejection(
    history: dict[str, Any], state: dict[str, Any], code: str
) -> tuple[bytes, int]:
    return output(history, state, code), 2


def reduce_history(history: dict[str, Any]) -> tuple[bytes, int]:
    state: dict[str, Any] = {
        "events": [],
        "root": 0,
        "standing": {},
        "submissions": [],
        "verifications": [],
        "versions": history["initial_versions"],
    }
    for record in history["records"]:
        kind = record["kind"]
        if kind == "submission":
            if not record["authenticated"]:
                continue
            state["submissions"].append((record["claim"], record["scope"]))
            state["standing"].setdefault(record["claim"], "unassessed")
            state["root"] += 1
            continue
        if kind == "verification":
            if (record["claim"], record["scope"]) not in state["submissions"]:
                continue
            state["verifications"].append((record["claim"], record["outcome"]))
            state["root"] += 1
            continue

        decision_id = record["id"]
        if record["repository"] != history["repository"]:
            return rejection(history, state, "wrong_repository")
        if record["performer"] not in history["authorized_performers"]:
            return rejection(history, state, "unauthorized")
        if record["authority_label"] != record["performer"]:
            return rejection(history, state, "misattributed")
        if record["expected_root"] != state["root"]:
            return rejection(history, state, "stale_root")
        if any(
            state["versions"].get(resource) != version
            for resource, version in record["read_set"].items()
        ):
            return rejection(history, state, "stale_read_set")

        action = record["action"]
        action_kind = action["kind"]
        if action_kind == "accept":
            claim = action["claim"]
            eligible = any(item[0] == claim for item in state["submissions"]) and any(
                item == (claim, "pass") for item in state["verifications"]
            )
        elif action_kind == "reject":
            claim = action["claim"]
            eligible = any(item[0] == claim for item in state["submissions"])
        else:
            replacement = action["replacement"]
            eligible = any(
                item[0] == replacement for item in state["submissions"]
            ) and any(item == (replacement, "pass") for item in state["verifications"])
        if not eligible:
            return rejection(history, state, "ineligible")

        if action_kind == "correct":
            prior = action["prior_decision"]
            predecessor = action["predecessor"]
            valid_reference = (
                any(
                    event["decision_id"] == prior
                    and event["repository"] == record["repository"]
                    and event["action"] == {"kind": "accept", "claim": predecessor}
                    for event in state["events"]
                )
                and state["standing"].get(predecessor) == "accepted"
            )
            if not valid_reference:
                return rejection(history, state, "invalid_correction_reference")

        if action_kind == "accept":
            state["standing"][action["claim"]] = "accepted"
        elif action_kind == "correct":
            state["standing"][action["predecessor"]] = "superseded"
            state["standing"][action["replacement"]] = "accepted"
        event = {
            "action": action,
            "authority_label": record["authority_label"],
            "decision_id": decision_id,
            "performer": record["performer"],
            "repository": record["repository"],
        }
        state["events"].append(event)
        state["root"] += 1
    return output(history, state, None), 0


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: reducer.py HISTORY.json", file=sys.stderr)
        return 64
    try:
        value = json.loads(Path(sys.argv[1]).read_text(encoding="utf-8"))
        history = validate_history(value)
        payload, code = reduce_history(history)
    except (OSError, json.JSONDecodeError, FormatError) as error:
        payload = (
            json.dumps(
                {"code": "invalid_format", "format": REJECTION_FORMAT},
                sort_keys=True,
                separators=(",", ":"),
            )
            + "\n"
        ).encode()
        print(error, file=sys.stderr)
        code = 2
    sys.stdout.buffer.write(payload)
    return code


if __name__ == "__main__":
    raise SystemExit(main())

File: paper/theory-of-standing/reducers/javascript/reducer.mjs
#!/usr/bin/env node
/** Independent JavaScript reducer for the proof-history interchange format. */

import fs from "node:fs";

const INPUT_FORMAT = "theory-of-standing.proof-history.v1";
const RESULT_FORMAT = "theory-of-standing.proof-result.v2";
const INVALID_FORMAT = "theory-of-standing.proof-invalid.v1";
const STANDING = new Set(["accepted", "unassessed", "superseded", "retracted"]);

class FormatError extends Error {}

function fail(message) {
  throw new FormatError(message);
}

function isObject(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function exact(value, keys, where) {
  if (!isObject(value)) fail(`${where}: expected object`);
  const actual = Object.keys(value).sort();
  const expected = [...keys].sort();
  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    fail(`${where}: expected exact keys ${JSON.stringify(expected)}`);
  }
  return value;
}

function nat(value, where) {
  if (!Number.isSafeInteger(value) || value < 0) {
    fail(`${where}: expected nonnegative safe integer`);
  }
  return value;
}

function natMap(value, where) {
  if (!isObject(value)) fail(`${where}: expected object`);
  const result = new Map();
  for (const [key, item] of Object.entries(value)) {
    if (!/^(0|[1-9][0-9]*)$/.test(key)) fail(`${where}: invalid resource id`);
    const resource = Number(key);
    nat(resource, `${where}: resource id`);
    result.set(resource, nat(item, `${where}.${key}`));
  }
  return result;
}

function validateAction(value, where) {
  if (!isObject(value) || typeof value.kind !== "string") fail(`${where}: invalid action`);
  if (value.kind === "accept" || value.kind === "reject") {
    exact(value, ["kind", "claim"], where);
    nat(value.claim, `${where}.claim`);
  } else if (value.kind === "correct") {
    exact(value, ["kind", "prior_decision", "predecessor", "replacement"], where);
    nat(value.prior_decision, `${where}.prior_decision`);
    nat(value.predecessor, `${where}.predecessor`);
    nat(value.replacement, `${where}.replacement`);
  } else {
    fail(`${where}: unsupported action`);
  }
  return value;
}

function validateHistory(value) {
  const history = exact(
    value,
    [
      "format",
      "repository",
      "authorized_performers",
      "initial_versions",
      "descriptive_dependencies",
      "records",
    ],
    "history",
  );
  if (history.format !== INPUT_FORMAT) fail("history.format: unsupported format");
  nat(history.repository, "history.repository");
  if (!Array.isArray(history.authorized_performers)) {
    fail("history.authorized_performers: expected array");
  }
  history.authorized_performers.forEach((actor, index) =>
    nat(actor, `history.authorized_performers[${index}]`),
  );
  const canonicalActors = [...new Set(history.authorized_performers)].sort((a, b) => a - b);
  if (JSON.stringify(canonicalActors) !== JSON.stringify(history.authorized_performers)) {
    fail("history.authorized_performers: expected sorted unique values");
  }
  history.initial_versions = natMap(history.initial_versions, "history.initial_versions");
  if (!Array.isArray(history.descriptive_dependencies)) {
    fail("history.descriptive_dependencies: expected array");
  }
  history.descriptive_dependencies.forEach((dependency, index) => {
    exact(dependency, ["dependent", "depends_on"], `dependency[${index}]`);
    nat(dependency.dependent, `dependency[${index}].dependent`);
    nat(dependency.depends_on, `dependency[${index}].depends_on`);
  });
  if (!Array.isArray(history.records)) fail("history.records: expected array");
  const decisionIds = new Set();
  history.records.forEach((record, index) => {
    const where = `record[${index}]`;
    if (!isObject(record) || typeof record.kind !== "string") fail(`${where}: invalid record`);
    if (record.kind === "submission") {
      exact(record, ["kind", "claim", "producer", "scope", "authenticated"], where);
      nat(record.claim, `${where}.claim`);
      nat(record.producer, `${where}.producer`);
      nat(record.scope, `${where}.scope`);
      if (typeof record.authenticated !== "boolean") fail(`${where}.authenticated: expected boolean`);
    } else if (record.kind === "verification") {
      exact(record, ["kind", "claim", "scope", "property", "outcome"], where);
      nat(record.claim, `${where}.claim`);
      nat(record.scope, `${where}.scope`);
      nat(record.property, `${where}.property`);
      if (record.outcome !== "pass" && record.outcome !== "fail") {
        fail(`${where}.outcome: unsupported outcome`);
      }
    } else if (record.kind === "decision") {
      exact(
        record,
        ["kind", "id", "repository", "authority_label", "performer", "expected_root", "read_set", "action"],
        where,
      );
      for (const key of ["id", "repository", "authority_label", "performer", "expected_root"]) {
        nat(record[key], `${where}.${key}`);
      }
      if (decisionIds.has(record.id)) fail(`${where}.id: duplicate Decision id`);
      decisionIds.add(record.id);
      record.read_set = natMap(record.read_set, `${where}.read_set`);
      validateAction(record.action, `${where}.action`);
    } else {
      fail(`${where}: unsupported kind`);
    }
  });
  return history;
}

function stableStringify(value) {
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(",")}]`;
  if (isObject(value)) {
    return `{${Object.keys(value)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${stableStringify(value[key])}`)
      .join(",")}}`;
  }
  return JSON.stringify(value);
}

function projection(history, state) {
  const corrected = new Set(
    state.events
      .filter((event) => event.action.kind === "correct")
      .map((event) => event.action.predecessor),
  );
  return [...state.standing.keys()]
    .sort((a, b) => a - b)
    .map((claim) => ({
      claim,
      status: history.descriptive_dependencies.some(
        (dependency) => dependency.dependent === claim && corrected.has(dependency.depends_on),
      )
        ? "needs_reassessment"
        : "unaffected",
    }));
}

function emit(history, state) {
  const standing = [...state.standing.entries()]
    .sort(([left], [right]) => left - right)
    .map(([claim, status]) => ({ claim, status }));
  if (standing.some((item) => !STANDING.has(item.status))) throw new Error("noncanonical Standing");
  const result = {
    events: state.events,
    format: RESULT_FORMAT,
    reassessment: projection(history, state),
    rejections: state.rejections,
    repository: history.repository,
    root: state.root,
    standing,
  };
  return `${stableStringify(result)}\n`;
}

function reduceHistory(history) {
  const state = {
    events: [],
    rejections: [],
    root: 0,
    standing: new Map(),
    submissions: [],
    verifications: [],
    versions: history.initial_versions,
  };
  for (let recordIndex = 0; recordIndex < history.records.length; recordIndex += 1) {
    const record = history.records[recordIndex];
    if (record.kind === "submission") {
      if (!record.authenticated) continue;
      state.submissions.push([record.claim, record.scope]);
      if (!state.standing.has(record.claim)) state.standing.set(record.claim, "unassessed");
      state.root += 1;
      continue;
    }
    if (record.kind === "verification") {
      if (!state.submissions.some(([claim, scope]) => claim === record.claim && scope === record.scope)) {
        continue;
      }
      state.verifications.push([record.claim, record.outcome]);
      state.root += 1;
      continue;
    }

    let code = null;
    if (record.repository !== history.repository) code = "wrong_repository";
    else if (!history.authorized_performers.includes(record.performer)) code = "unauthorized";
    else if (record.authority_label !== record.performer) code = "misattributed";
    else if (record.expected_root !== state.root) code = "stale_root";
    else {
      for (const [resource, version] of record.read_set) {
        if (state.versions.get(resource) !== version) {
          code = "stale_read_set";
          break;
        }
      }
    }

    const action = record.action;
    let eligible;
    if (action.kind === "accept") {
      eligible = state.submissions.some(([claim]) => claim === action.claim)
        && state.verifications.some(([claim, outcome]) => claim === action.claim && outcome === "pass");
    } else if (action.kind === "reject") {
      eligible = state.submissions.some(([claim]) => claim === action.claim);
    } else {
      eligible = state.submissions.some(([claim]) => claim === action.replacement)
        && state.verifications.some(
          ([claim, outcome]) => claim === action.replacement && outcome === "pass",
        );
    }
    if (code === null && !eligible) code = "ineligible";

    if (code === null && action.kind === "correct") {
      const validReference = state.events.some(
        (event) => event.decision_id === action.prior_decision
          && event.repository === record.repository
          && event.action.kind === "accept"
          && event.action.claim === action.predecessor,
      ) && state.standing.get(action.predecessor) === "accepted";
      if (!validReference) code = "invalid_correction_reference";
    }

    if (code !== null) {
      state.rejections.push({ code, record_index: recordIndex });
      continue;
    }

    if (action.kind === "accept") state.standing.set(action.claim, "accepted");
    if (action.kind === "correct") {
      state.standing.set(action.predecessor, "superseded");
      state.standing.set(action.replacement, "accepted");
    }
    state.events.push({
      action,
      authority_label: record.authority_label,
      decision_id: record.id,
      performer: record.performer,
      repository: record.repository,
    });
    state.root += 1;
  }
  return { bytes: emit(history, state), exitCode: 0 };
}

function main() {
  if (process.argv.length !== 3) {
    process.stderr.write("usage: reducer.mjs HISTORY.json\n");
    return 64;
  }
  try {
    const history = validateHistory(JSON.parse(fs.readFileSync(process.argv[2], "utf8")));
    const result = reduceHistory(history);
    process.stdout.write(result.bytes);
    return result.exitCode;
  } catch (error) {
    if (!(error instanceof FormatError) && !(error instanceof SyntaxError)) throw error;
    process.stderr.write(`${error.message}\n`);
    process.stdout.write(`${stableStringify({ code: "invalid_format", format: INVALID_FORMAT })}\n`);
    return 2;
  }
}

process.exitCode = main();
